### PR TITLE
json implemented in zfs CLI

### PR DESCRIPTION
**Pull request by goulvenriou**
Hi, in response to  the pull request https://github.com/zfsonlinux/zfs/pull/3483 Alyseo's team has refactored the code to deliver single commit including all the jsonify code, the json output has been implements in all CLI commands.
Regards.
Goulven

### CODE REVIEW

**Comment by behlendorf** [comment]
Awesome, thanks for squashing the patch stack it will make it much easier to review.

**Comment by ryao** [comment]
@goulvenriou @gwilson told me that you have libzfs_core extensions as part of the JSON changes. Where can I find those?

**Comment by goulvenriou** [comment]
@ryao  , Hi ,  the libzfs_core isn't affected by the json implementation. Regards Goulven

**Comment by hadees** [comment]
Awesome! I've really been wanting to write a web interface for my box and this will make it way easier. :+1:

**Comment by yada** [comment]
@hadees Was one of the main goal : we think it will facilitate many new community projects based on top of openzfs ;-) Happy if it help you and please test, review, comment or create bug report if you find any. Thanks for your support.

**Comment by hadees** [comment]
@yada is this live yet? How do I use it?

**Comment by yada** [comment]
@hadees The branch has been updated by our team (it is up to date now) and you check this readme file  if you want to build : https://github.com/Alyseo/zfs/blob/json-0.6.5/json/install.md

**Comment by fractalram** [comment]
Is this part of the newly release zfs-0.6.5.4?

**Comment by yada** [comment]
@fractalram Still not, we will push again a new and fresh rebase to help core dev review and hope it will happen soon ;-) @behlendorf may have more info on timeframe ?

**Comment by yada** [comment]
New rebase done, hope it helps ;-)

**Comment by yada** [comment]
Hi,

Been a while I did not reply to this thread but lots of progress has been done in the mean time.

We understand it is a pretty disruptive append to the code base but at least it is not IMHO introduce disruptive changes.

We are closed to the final step now, all commands have been implemented using two formats (json and ldjson), schemas descriptions are online (including orderly), we regularly push new rebase to help code review... So now we just need some love :) or in other words, we need to know if you guys (OpenZFS community) agree and get buy on the fundamental approach and give their approval for merging our code ?

For infos (output and format...), see :
https://github.com/Alyseo/zfs/tree/json-0.6.5/json

BTW, we follow and like #3907 
We have different approaches but I strongly think that both can be very complementary and can co-exist since they serve slightly different roles. To help code review and these two importants pull requests, maybe we can help each others on code review ? @ryao let me know what you think about it and if it can helps ?

PS: Also posted to [OpenZFS Developer]

Regards,
Yacine

**Comment by goulvenriou** [comment]
new rebase Donne , your comments and reviews are welcome . the diff is available here : https://github.com/zfsonlinux/zfs/pull/3938/files
Regards,
Goulven

**Comment by yada** [comment]
The usual tests failed I guess, can you guys confirm we have nothing to fix on our side to improve this : @behlendorf ?

**Comment by behlendorf** [comment]
@yada all the tests normally pass.  It looks like your changes have accidentally introduced at least one change in behavior.  It seems filebench is failing because the `zfs set compression` command is failing, see the [CentOS 7 test log](http://build.zfsonlinux.org/builders/CentOS%206.7%20x86_64%20%28TEST%29/builds/600/steps/shell_9/logs/stdio) as an example.

```
sudo -E zfs set compression=lz4 tank/filebench
missing property=value argument(s)
usage:
	set [-jJ] <property=value> ... <filesystem|volume|snapshot> ...

The following properties are supported:
...
```

**Comment by yada** [comment]
@behlendorf : thanks for the info and issue fixed, all passed now and branch is up to date so happy review ;-)

**Comment by behlendorf** [comment]
@yada how goes the process of adding test cases to verify the json output?

**Comment by yada** [comment]
@behlendorf Sorry, team been busy on big project... will find some spare time to work on it asap.

**Comment by ptx0** [comment]
I don't think these changes should be merged until others review the proposed schema. I have at least one major issue. As well, the proposed schema aren't even complete.. many do not even have examples as to what it will look like.

The JSON changes should make it easier to use ZFS CLI programmatically, but they currently do not. I would prefer there be zero JSON output, rather than something that's not useful.

**Comment by yada** [comment]
Can you describe and explain your major issue !?
All schema are complete for ages... See :
https://github.com/Alyseo/zfs/blob/json-0.6.5/json/STATUS.md

What is missing from your point of view ?
Will love to understand your latest sentence and why current work done is not useful...

**Comment by denji** [comment]
https://github.com/Juniper/libxo

**Comment by goulvenriou** [comment]
@denji  
the choice of the library was made ​​during exchange on the mailing list zfs before implementation, you can see the archive here : https://www.mail-archive.com/developer@open-zfs.org/msg00880.html ,
 we have indeed chosen to use the libnvpair , modified by Joyent
```
FYI : 
- Long term, we'd like the kernel to do more of the property processing, so this could eventually turn into simply:
     - do an ioctl to get the requested properties of the requested datasets, returning an nvlist
     - print nvlist as json (which has already been implemented, e.g. by Joyent)

```

File: json/schemav1.0/schema/schema-zfs-userspace.json [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        },
+        "stdout": {
+            "type":"array",
+            "description": "standard output of command",
+            "minitems": "0",
+            "required":true,
+            "items":
+                {
+                    "type":"object",
+                    "type": "type of volume",
```
Comment (wspeirs) [inline json/schemav1.0/schema/schema-zfs-userspace.json]: `type` used twice... same as in schema-zfs-list.json

File: json/schemav1.0/schema/schema-zfs-upgrade.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+            "required":false
+        },
+        "stdout": {
+            "type":"array",
+            "name": "stdout",
+            "description": "standard output of command",
+            "minitems": "0",
+            "required":true
+        }
+    }
```
Comment (wspeirs) [inline json/schemav1.0/schema/schema-zfs-upgrade.json]: Missing closing brace

File: json/schemav1.0/schema/schema-zfs-snapshot.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            "name": "stdout",
+            "description": "standard output of command",
+            "minitems": "0",
+            "required":true
+        }
+    }
+}
+
+
+```
```
Comment (wspeirs) [inline json/schemav1.0/schema/schema-zfs-snapshot.json]: Erroneous markdown

File: json/schemav1.0/schema/schema-zfs-mount.json [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+                            "required":false
+                        },
+                        "mountpoint": {
+                            "type":"string",
+                            "description": "mountpoint of: volume in filesystem",
+                            "required":false
+                        }
+        }
+    }
+}
```
Comment (wspeirs) [inline json/schemav1.0/schema/schema-zfs-mount.json]: Missing closing brace

File: json/schemav1.0/schema/schema-zfs-list.json [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                            "required":false
+                        }
+                    }
+                }
+        }
+    }
+}
+
+
+```
```
Comment (wspeirs) [inline json/schemav1.0/schema/schema-zfs-list.json]: Erroneous markdown

File: json/schemav1.0/schema/schema-zfs-groupspace.json [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        },
+        "stdout": {
+            "type":"array",
+            "description": "standard output of command",
+            "minitems": "0",
+            "required":true,
+            "items":
+                {
+                    "type":"object",
+                    "type": "type of volume",
```
Comment (wspeirs) [inline json/schemav1.0/schema/schema-zfs-groupspace.json]: `type` seems to be a duplicate key here, and overrides the schema restriction. Should "type of volume" be a property instead?

File: json/schemav1.0/schema/schema-zfs-diff.json [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        },
+        "stdout": {
+            "type":"array",
+            "description": "standard output of command",
+            "minitems": "0",
+            "required":true,
+            "items":
+                {
+                    "type":"object",
+                    "type": "type of volume",
```
Comment (wspeirs) [inline json/schemav1.0/schema/schema-zfs-diff.json]: `type` seems to be a duplicate key here, and overrides the schema restriction. Should "type of volume" be a property instead?

File: json/schemav1.0/schema/schema-zfs-destroy.json [hunk 8]
```diff
@@ -0,0 +1,27 @@
+zfs destroy  -J :
```
Comment (wspeirs) [inline json/schemav1.0/schema/schema-zfs-destroy.json]: Erroneous sring

File: json/schemav1.0/schema/schema-zfs-create.json [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+            "name": "stdout",
+            "description": "standard output of command",
+            "minitems": "0",
+            "required":true
+        }
+    }
+}
+
+
+```
```
Comment (wspeirs) [inline json/schemav1.0/schema/schema-zfs-create.json]: Erroneous markdown

**Comment by wspeirs** [comment]
@yada I also believe there are errors in your schema files compared to what is generated. Using `zfs list -J` on a system with no ZFS volumes yields the following output:

```
{"cmd":"zfs list","schema_version":"1.0","stdout":[],"stderr":{"error":"no datasets available"}}
```

However, according to the schema, `stderr` should be a string:

```
        "stderr": {
            "type":"string",
            "title": "stderr",
            "description": "error output of command",
            "required":true
        },
```

I imagine this is broken in a few places...

Also, why did you use [Orderly](https://github.com/lloyd/orderly) (the orderly-json.org site seems to be down) instead of straight [JSON schema](http://json-schema.org/)? I *think* the next step to creating tests would be to run commands and verify the output against the schema files, but currently too much is broken for this to work. Happy to help, but I need to better understand the design decisions here.

**Comment by alek-p** [comment]
Looks like this PR has been inactive for a while, @goulvenriou, @yada what's the status on this?
We'd like to see this integrated so I can offer assistance to move this along.

**Comment by kpande** [comment]
yes, there were a number of inconsistencies and other issues with the schema.

**Comment by yada** [comment]
@alek-p : Yep sorry guys but new job and no ressources anymore to move forward on this project. 
I think this feature will be great for the openzfs community and hope some folks can take over...

**Comment by sevagh** [comment]
Hello. I've had to write some ugly Python to parse the output of `zfs status`, and searched for JSON output. I'm interested in continuing the work on this PR. Should I continue from the last commit on this PR, or is it better to start over?

**Comment by ptx0** [comment]
either one is going to be ugly, may as well just make a python kit for ZFS.

**Comment by behlendorf** [comment]
There was work done by ClusterHQ https://github.com/ClusterHQ/pyzfs which is worth looking at.

**Comment by alek-p** [comment]
We've made some progress on JSON output support for `zfs` cmd at datto. What we've done is piggyback JSON output support on top of channel programs (#6558). This way the JSON output support is targeted to scripting use cases and is easily maintainable since it really only touches one function (`zfs_do_channel_program()`).
Of course, this only adds the ability to get JSON output for zfs commands, we're still thinking about how to cleanly do this for zpool commands and welcome collaboration on that.

Since channel programmes got merged recently, I've dug out the JSON patch and am rebasing it, plus adding tests. Once that's done it will be put out for review in ZoL.

**Comment by yzgyyang** [comment]
> We've made some progress on JSON output support for `zfs` cmd at datto. What we've done is piggyback JSON output support on top of channel programs (#6558). This way the JSON output support is targeted to scripting use cases and is easily maintainable since it really only touches one function (`zfs_do_channel_program()`).
> Of course, this only adds the ability to get JSON output for zfs commands, we're still thinking about how to cleanly do this for zpool commands and welcome collaboration on that.
> 
> Since channel programmes got merged recently, I've dug out the JSON patch and am rebasing it, plus adding tests. Once that's done it will be put out for review in ZoL.

Any news regarding `zpool` channel programs?

**Comment by alek-p** [comment]
> > We've made some progress on JSON output support for `zfs` cmd at datto. What we've done is piggyback JSON output support on top of channel programs (#6558). This way the JSON output support is targeted to scripting use cases and is easily maintainable since it really only touches one function (`zfs_do_channel_program()`).
> > Of course, this only adds the ability to get JSON output for zfs commands, we're still thinking about how to cleanly do this for zpool commands and welcome collaboration on that.
> > Since channel programmes got merged recently, I've dug out the JSON patch and am rebasing it, plus adding tests. Once that's done it will be put out for review in ZoL.
> 
> Any news regarding `zpool` channel programs?

yes it was implemented here https://github.com/openzfs/zfs/pull/7281

**Pull request closed without merging**